models/user.go

```diff
@@ -1402,35 +1406,144 @@ func SyncExternalUsers() {
+							key, err := AddPublicKey(usr.ID, LDAPPublicSSHKeyName, LDAPPublicSSHKey, s.ID)
+							if err != nil {
+								log.Error(4, "SyncExternalUsers[%s]: Error adding LDAP Public SSH Key for user %s: %v [%v]", s.Name, su.Username, key, err)
+							} else {
+								sshKeysNeedUpdate = true
+							}
+						}
 					}
 				} else if updateExisting {
 					existingUsers = append(existingUsers, usr.ID)
-					// Check if user data has changed
-					if (len(s.LDAP().AdminFilter) > 0 && usr.IsAdmin != su.IsAdmin) ||
-						strings.ToLower(usr.Email) != strings.ToLower(su.Mail) ||
-						usr.FullName != fullName ||
-						!usr.IsActive {
-
-						log.Trace("SyncExternalUsers[%s]: Updating user %s", s.Name, usr.Name)
-
-						usr.FullName = fullName
-						usr.Email = su.Mail
-						// Change existing admin flag only if AdminFilter option is set
-						if len(s.LDAP().AdminFilter) > 0 {
-							usr.IsAdmin = su.IsAdmin
-						}
-						usr.IsActive = true
 
-						err = UpdateUser(usr)
+					if isAttributeSSHPublicKeySet {
+						log.Trace("SyncExternalUsers[%s]: Handling user %s", s.Name, usr.Name)
+
+						// Get Public Keys from DB with LDAP source
+						var giteaKeys []string
+						keys, err := ListPublicKeys(usr.ID)
+
 						if err != nil {
-							log.Error(4, "SyncExternalUsers[%s]: Error updating user %s: %v", s.Name, usr.Name, err)
+							log.Error(4, "SyncExternalUsers[%s]: Error listing LDAP Public SSH Keys for user %s: %v", s.Name, usr.Name, err)
+						}
+						for _, v := range keys {
+							// If key was synced from LDAP
+							if v.LoginSourceID > 0 {
+								giteaKeys = append(giteaKeys, v.OmitEmail())
+							}
+						}
+
+						// Get Public Keys from LDAP and skip duplicate keys
+						var ldapKeys []string
+						for _, v := range su.SSHPublicKey {
+							ldapKey := strings.Join(strings.Split(v, " ")[:2], " ")
+							// Check if value is already in slice, else add it
+							i := sort.SearchStrings(ldapKeys, ldapKey)
+							if i < len(ldapKeys) && ldapKeys[i] == ldapKey {
+							} else {
+								ldapKeys = append(ldapKeys, ldapKey)
+							}
+						}
+
+						// Sort key slices
+						sort.Strings(ldapKeys)
+						sort.Strings(giteaKeys)
+
+						// Check if sync is needed
+						var giteaKeysToDelete []string
+						if reflect.DeepEqual(giteaKeys, ldapKeys) {
+							log.Trace("SyncExternalUsers[%s]: Public Keys are already in sync for %s (LDAP:%v/DB:%v)", s.Name, usr.Name, len(ldapKeys), len(giteaKeys))
+							continue
+						} else {
+							log.Trace("SyncExternalUsers[%s]: Public Key needs update for user %s (LDAP:%v/DB:%v)", s.Name, usr.Name, len(ldapKeys), len(giteaKeys))
+							// Delete giteaKeys that doesn't exist in ldapKeys and has been synced from LDAP earlier
+							for _, giteaKey := range giteaKeys {
+								i := sort.SearchStrings(ldapKeys, giteaKey)
+								if i < len(ldapKeys) && ldapKeys[i] == giteaKey {
+									log.Trace("SyncExternalUsers[%s]: Key exists in LDAP: %s", s.Name, giteaKey)
```

Comment:
Please try to split this into separate functions. All these nested if-for-if-for's are just a pain to read

![](http://i.imgur.com/BtjZedW.jpg)